router/io-spec-non-proven-lemmas.gobra

```diff
@@ -166,6 +166,68 @@ pure func (p* scionPacketProcessor) GetIsXoverSpec(ub []byte) bool {
 		p.path.GetIsXoverSpec(ubPath)
 }
 
+ghost
+opaque
+requires acc(&p.d, R55) && acc(p.d.Mem(), _)
+requires acc(&p.ingressID, R55)
+requires len(pkt.CurrSeg.Future) > 0
+decreases
+pure func (p *scionPacketProcessor) NoBouncingPkt(pkt io.IO_pkt2) bool {
+	return let currseg := pkt.CurrSeg in
+		let OptEgressID := (currseg.ConsDir ? currseg.Future[0].EgIF2 : currseg.Future[0].InIF2) in
+		let egressID := OptEgressID != none[io.IO_ifs] ? uint16(get(OptEgressID)) : 0 in
+		((egressID in p.d.getDomExternal()) || p.ingressID != 0)
+}
+
+ghost
+requires acc(&p.d, R55) && acc(p.d.Mem(), _)
+requires acc(&p.ingressID, R55)
+requires len(pkt.CurrSeg.Future) > 0
+requires AbsEgressInterfaceConstraint(pkt, path.ifsToIO_ifs(egressID))
+requires acc(&p.d.external, _)
+requires p.d.external == nil ==> p.ingressID != 0
+requires p.d.external != nil ==> acc(p.d.external, _)
+requires p.d.external != nil ==>
+	(egressID in domain(p.d.external)) || p.ingressID != 0
```

Comment:
Also, is there a reason to explicitly pass permissions to `acc(p.d.external, _)`  and to access `p.d.external` directly instead of calling `p.d.getDomExternal()`?

Reply:
> I guess we can replace the two following preconditions
> 
> ```
> requires p.d.external == nil ==> p.ingressID != 0
> ...
> requires p.d.external != nil ==>
>  	(egressID in domain(p.d.external)) || p.ingressID != 0
> ```
> 
> by
> 
> ```
> (egressID in domain(p.d.external)) || p.ingressID != 0
> ```

Yes you are right, I confused myself with the permissions there. 



> Also, is there a reason to explicitly pass permissions to `acc(p.d.external, _)` and to access `p.d.external` directly instead of calling `p.d.getDomExternal()`?

`p.d.getDomExternal()` unfolds the dataplane's `Mem()` predicate and is not `opaque` at the moment. If we want to use it directly we would have to refactor a bit. Should I do it?

Reply:
if you have time, then we could try to make this opaque and unfold it only where it is needed. It might have the nice side effect of actually simplifying some proof obligations, but I don't have a good feeling for whether it can help a lot right now. If you decide to stick with the current solution, at least mention the reason why we are passing the permission to `d.external` explicitly

Reply:
I think it was worth doing. It simplified some proof obligations and sped up verification times of some functions.